crates/apollo-router/src/configuration/otlp/mod.rs

```diff
@@ -80,6 +80,37 @@ impl ExportConfig {
     }
 }
 
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct EndpointUrl {
+    #[serde(deserialize_with = "endpoint_url")]
+    endpoint: Url,
+}
+
+fn endpoint_url<'de, D>(deserializer: D) -> Result<Url, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let mut buf = String::deserialize(deserializer)?;
+
+    // support the case of a IP:port endpoint
+    if buf.parse::<std::net::SocketAddr>().is_ok() {
+        buf = format!("https://{}", buf);
+    }
+
+    let mut url = Url::parse(&buf).map_err(serde::de::Error::custom)?;
+
+    // support the case of 'collector:4317' where url parses 'collector'
+    // as the scheme instead of the host
+    if url.host().is_none() && (url.scheme() != "http" || url.scheme() != "https") {
+        buf = format!("https://{}", buf);
+
+        url = Url::parse(&buf).map_err(serde::de::Error::custom)?;
+    }
```

Comment:
Why not use set_scheme?

Reply:
because we would end up with the url "https://4317" instead of "https://collector:4317". The url crate recognizes "collector" as the scheme in "collector:4317"